src/lib.rs

```diff
@@ -788,42 +788,52 @@ impl<T: Trait> Module<T> {
                 consensus_height,
             } => {
                 ensure!(
-                    ClientStates::contains_key(&client_identifier),
+                    ClientStates::contains_key(&client_id),
                     "Client not found"
                 );
-                ensure!(
-                    !Connections::contains_key(&desired_identifier),
-                    "Connection identifier already exists"
-                );
+
+                let mut new_connection_end;
+                if Connections::contains_key(&connection_id) {
+                    let old_conn_end = Connections::get(&connection_id);
+                    let state_is_consistent = old_conn_end.state.eq(&ConnectionState::Init)
+                        && old_conn_end.counterparty_connection_id.eq(&counterparty_connection_id)
+                        && old_conn_end.counterparty_client_id.eq(&counterparty_client_id);
+
+                    ensure!(state_is_consistent, "Local connection corrupted!");
```

Comment:
I think if the state is inconsistent here, this function will return, and the following code will not have a chance to execute.